src/nnet3/nnet-simple-component.cc

```diff
@@ -668,12 +718,99 @@ void ClipGradientComponent::Backprop(const std::string &debug_info,
     }
+
+    to_update->num_backproped_ += 1;
+    RepairGradients(debug_info, in_value, in_deriv, to_update);
   }
 }
 
+void ClipGradientComponent::RepairGradients(
+    const std::string &debug_info,
+    const CuMatrixBase<BaseFloat> &in_value,
+    CuMatrixBase<BaseFloat> *in_deriv, ClipGradientComponent *to_update) const {
+  KALDI_ASSERT(to_update != NULL);
+  BaseFloat default_clipped_proportion_threshold = 0.01;
+  BaseFloat default_clipping_to_value = 0.0;
+
+  // we use this 'repair_probability' (hardcoded for now) to limit
+  // this code to running on about half of the minibatches.
+  BaseFloat repair_probability = 0.5;
+  if (self_repair_clipped_proportion_threshold_ >= 1.0 ||
+      self_repair_scale_ == 0.0 || count_ == 0 ||
+      RandUniform() > repair_probability)
+    return;
+
+  BaseFloat unset = kUnsetValue; // -1000.0
+
+  BaseFloat clipped_proportion_threshold =
+      (self_repair_clipped_proportion_threshold_ == unset ?
+       default_clipped_proportion_threshold :
+       self_repair_clipped_proportion_threshold_);
+  BaseFloat clipping_to_value =
+      (self_repair_clipping_to_value_ == unset ?
+       default_clipping_to_value :
+       self_repair_clipping_to_value_);
+  KALDI_ASSERT(clipping_to_value >= 0.0 && self_repair_scale_ > 0.0);
+
+  BaseFloat clipped_proportion =
+    (count_ > 0 ? static_cast<BaseFloat>(num_clipped_) / count_ : 0);
+  // in-deriv would be modified only when clipped_proportion exceeds the
+  // threshold
+  if (clipped_proportion <= clipped_proportion_threshold)
+    return;
+
+  to_update->num_self_repaired_ += 1;
+  if (to_update->debug_info_ == "") //debug
+    to_update->debug_info_ = debug_info; //debug
+  if (to_update->num_self_repaired_ == 1)
+    KALDI_LOG << "ClipGradientComponent(node_name=" << debug_info
+              << ")'s self-repair was activated as the first time at the "
+              << num_backproped_
+              << "-th call of Backprop() in this training job.";
+
+  // sign_mat = sign(in_value), i.e.,
+  // An element in sign_mat is 1 if its corresponding element in in_value > 0,
+  // or -1 otherwise
+  CuMatrix<BaseFloat> sign_mat(in_value);
+  sign_mat.ApplyHeaviside();
+  sign_mat.Scale(2.0);
+  sign_mat.Add(-1.0);
+
+  // repair_mat =
+  // floor(abs(in_value) - clipping_to_value, 0) .* sign(in_value)
+  CuMatrix<BaseFloat> repair_mat(in_value);
+  repair_mat.ApplyPowAbs(1.0);
+  repair_mat.Add(-clipping_to_value);
+  repair_mat.ApplyFloor(0.0);
+  repair_mat.MulElements(sign_mat);
+
+  // magnitude =
+  // self_repair_scale_ * clipped_proportion * average norm of in-deriv
+  CuVector<BaseFloat> in_deriv_norm_vec(in_deriv->NumRows());
+  in_deriv_norm_vec.AddDiagMat2(1.0, *in_deriv, kNoTrans, 0.0);
+  in_deriv_norm_vec.ApplyPow(0.5);
+  double in_deriv_norm_sum = in_deriv_norm_vec.Sum();
+  BaseFloat magnitude = self_repair_scale_ * clipped_proportion *
+                        (in_deriv_norm_sum / in_deriv_norm_vec.Dim());
```

Comment:
Yes but you are then applying that magnitude to each individual row of the
matrix.  What I am proposing instead is that you just scale the matrix so
that on average the rows have a certain magnitude.  It would be easier to
not think about magnitudes per row, but magnitudes at a whole-matrix level.

On Tue, May 31, 2016 at 6:13 PM, Yiming Wang <notifications@github.com>
wrote:

> In src/nnet3/nnet-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/803#discussion_r65272751>:
>
> > +  // repair_mat =
> > +  // floor(abs(in_value) - clipping_to_value, 0) .* sign(in_value)
> > +  CuMatrix<BaseFloat> repair_mat(in_value);
> > +  repair_mat.ApplyPowAbs(1.0);
> > +  repair_mat.Add(-clipping_to_value);
> > +  repair_mat.ApplyFloor(0.0);
> > +  repair_mat.MulElements(sign_mat);
> > +
> > +  // magnitude =
> > +  // self_repair_scale_ * clipped_proportion * average norm of in-deriv
> > +  CuVector<BaseFloat> in_deriv_norm_vec(in_deriv->NumRows());
> > +  in_deriv_norm_vec.AddDiagMat2(1.0, *in_deriv, kNoTrans, 0.0);
> > +  in_deriv_norm_vec.ApplyPow(0.5);
> > +  double in_deriv_norm_sum = in_deriv_norm_vec.Sum();
> > +  BaseFloat magnitude = self_repair_scale_ * clipped_proportion *
> > +                        (in_deriv_norm_sum / in_deriv_norm_vec.Dim());
>
> Here in_deriv_norm_sum is the sum of the norm of each row of in_deriv, so
> the magnitude is already set at the whole matrix level (average of the norm
> of each row of the matrix in_deriv).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/803/files/149839bb422bad0bd9311e7d3a74a3f35967e47a#r65272751>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADJVu8TN0PuJBHJ8-6-OQ8ByutX8XhCtks5qHLJ_gaJpZM4IkDwb>
> .
>